### DISCUSSION TITLE
Building examples

### DISCUSSION THREAD

**Discussion by yanboti** [randombit/botan#3789] [Q&A] [answered]
Hi, I would like to selectively choose which examples to compile. In particular, I would like to build only Kyber.
Furthermore, I would like to build only modules that I require. As such, I'm trying to configure with the following command:
./configure.py 
-cc=gcc -cxxflags=-std=c++20 
--with-os-features=getentropy 
--with-external-includedir=/usr/include 
--with-external-includedir=/usr/include/x86-64-linux-gnu
--with-external-libdir=/usr/lib
--with-external-libdir=/usr/lib/x86-64-linux-gnu
--minimized-build
--enable-modules=kyber
--enable-modules=keccak
--enable-system_rng
--build-targets=static
--build-targets=shared
--build-targets=examples

But this runs into problems examples will try to build all the examples in the /src/examples folder. What should I do? I've tried renaming the example source files too, but that created other warnings with dependencies in the makefile.

**Comment by securitykernel** [comment, marked as answer]
I am sorry you experience issues with the build system. The examples are currently all-or-nothing, you can not selectively enable or disable single or multiple examples. Some require boost, so all currently require boost, as you noted that in #3790. With #3787 merged yesterday, `configure.py` checks that now and fails if boost isn't enabled with the examples.

What you could do instead, is do your minimized kyber-only build, and then compile the kyber example on your own, for example:

```bash
$ ./configure.py --enable-modules=kyber ... [all your additional options]
$ g++ kyber.cpp -o kyber -L. -lbotan-3 -Ibuild/include -std=c++20
```

**Reply by yanboti** [reply]
Thank you very much for your reply.
I've done this, but there are modules that are missing that results in a seg fault.
I think I need to enable more than just kyber, kdf, pubkey and system_rng. The line that fails is 
src/examples/kyber.cpp L19:  `Botan::PK_KEM_Encryptor enc(*pub_key, kdf);`
Do you know what other modules I might need to enable?

I actually compiled the example using
`g++ src/examples/kyber.cpp -o kyber -L. -lbotan-3 -Ibuild/include -std=c++20 -lpthread`

**Reply by randombit** [reply]
Instead of refusing to compile all of the examples in this case maybe we can introduce `#if defined(BOTAN_HAS_...)` checks and in the event that the example cannot be compiled or used, compiling instead something like

```
int main() {
  std::cerr << "This example is unavailable due to missing feature X";
  return 1;
}
```

**Reply by lieser** [reply]
@yanboti you are missing the `hkdf` and `sha2_64` module.

**Reply by randombit** [reply]
This looks like a bug - we should not segfault if the KDF is not available

**Reply by lieser** [reply]
> This looks like a bug - we should not segfault if the KDF is not available

Because of that I tried to reproduce it, but was unable. I think this was "just" an unhandled `LookupError` exception because of the `hkdf` and `sha2_64` module.

**Reply by randombit** [reply]
OK looks like we don't segfault, it just exits with an uncaught exception

**Reply by yanboti** [reply]
Hello good people of botan, you are right that I didn't seg fault and what happened was that an abort signal was detected. I also subsequently realised that I was missing hkdf and sha2 modules. Thank you for your help!
